clang/test/CodeGen/PowerPC/builtins-bcd-assist.c

```diff
@@ -0,0 +1,58 @@
+// NOTE: Assertions have been autogenerated by utils/update_cc_test_checks.py UTC_ARGS: --version 5
+// REQUIRES: powerpc-registered-target
+// RUN: %clang_cc1 -triple powerpc64le-unknown-linux -O2 -target-cpu pwr7 \
+// RUN:   -emit-llvm %s -o - | FileCheck %s
+// RUN: %clang_cc1 -triple powerpc64-unknown-aix -O2 -target-cpu pwr7 \
+// RUN:   -emit-llvm %s -o - | FileCheck %s
+// RUN: %clang_cc1 -triple powerpc-unknown-aix -O2 -target-cpu pwr7 \
+// RUN:   -emit-llvm %s -o - | FileCheck %s
+
+// CHECK-LABEL: define{{.*}} i64 @cdtbcd_test(i64
+// CHECK:         [[CONV:%.*]] = trunc i64 {{.*}} to i32
+// CHECK-NEXT:    [[TMP0:%.*]] = tail call i32 @llvm.ppc.cdtbcd(i32 [[CONV]])
+// CHECK-NEXT:    [[CONV1:%.*]] = zext i32 [[TMP0]] to i64
+// CHECK-NEXT:    ret i64 [[CONV1]]
+long long cdtbcd_test(long long ll) {
+    return __builtin_cdtbcd (ll);
```

Comment:
Hmm, for compatibility with GCC, maybe this is right. But this is not good as in clang we are also providing `__cdtbcd()` for 64 bit inputs. Using GCC's 32-bit version will cause sacrifice of accuracy. I am not sure, for this case, should we add at least a warning and remind the user the other one can be used?

Reply:
We are only providing `__xxx()` in 64bit for backward compat to XL.  Doc will be udpated to ask user to use `__builtin_xxx()` for 32bit when migrating to OpenXL.  GCC `__builtin_xxx()` only provide support for `unsigned int`.  Do we really want to direct user to a functiont that takes `signed long long` type?

Reply:
> Do we really want to direct user to a functiont that takes `signed long long` type?

Some users can benefit from the XL version of the functions even if they are not migrating from XL; however, they would need to be willing to have code that is specific to 64-bit IBM XL and Clang. The 64-bit versions of `__cbcdtd` and `__cdtbcd` operate on two 32-bit operands at the same time. The 64-bit version of `addg6s` operates on wider BCD operands.

Also, to be clear, we are only redirecting users of `__cbcdtd` and `__cdtbcd` in 32-bit mode to the GCC versions (where they might need to call the GCC built-in twice if they had a 64-bit input: once for the high 32-bits and once for the low 32-bits).

`__addg6s` was always 64-bit only for IBM XL, so there is no migration-from-IBM-XL scenario involving 32-bit `__addg6s`.

Reply:
> Do we really want to direct user to a function that takes signed long long type?

My comment above is for case when a user wants to call `cdtbcd` for a 64-bit input for 64-bit mode(for new codes). If the GCC compat one `__builtin_cdtbcd()` is chosen by accident(this is not rejected in this patch, although GCC builtin should only accept 32-bit input), the input will be trunc so the output will be not accurate(?).  But if the XLC compat one `__cdtbcd()` is used, I think the output is accurate?

Reply:
Why this patch doesn't reject 64-bit input for `__builtin_cdtbcd` like it rejects 32-bit input for `__cdtbcd`?

Reply:
> like it rejects 32-bit input for `__cdtbcd`

It does not do so though. It _accepts_ (and converts) 32-bit arguments (in the C/C++ code) for `__cdtbcd` (in 64-bit mode).